README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# rabi

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/rabi)](https://cran.r-project.org/package=rabi)

The goal of `rabi` is to facilitate the design and generation of color
(or symbol) codes that can be used to mark and identify individual
animals (specifically aimed for research projects). These codes can be
selected such that the IDs are robust to partial erasure: even if parts
of the code are lost, the entire identity of the animal can be
reconstructed Thus, animal subjects are not confused and no ambiguity is
introduced.

## Installation

You can install `rabi` from CRAN with:

``` r
install.packages("rabi")
```

You can also install `rabi` from Github with:

``` r
# install.packages("devtools")
devtools::install_github("andburch/rabi")
```

## Example

Although these methods and principles are applicable across taxa and
situations, we will demonstrate their usefulness on ants as an
illustration. A well-thought-out identification system is important in
social insect research because colonies are often composed of hundreds
to thousands of virtually identical workers. Additionally, their
tendency to groom themselves (and each other) creates a situation where
markings are especially susceptible to becoming detached and disrupting
identification.

Let’s say we have 100 ants we need to track and observe: we need to
*always* know their individual identities. We have five colors of paint
which we will apply as drops to locations on their backs. Their heads
and thoraxes will each receive a colored drop and then they will get two
drops on their abdomens (i.e. gasters). We want a coding scheme that
will still perfectly identify the ants even if they *all* lose one of
these marks.

``` r
total.length <- 4  #we have four places to put paint,
redundancy <- 1    #we want robustness to single erasures,
alphabet <- 5      #and we currently have five types of paint in stock
 #rs_IDs() is one of the functions that can generate codes
codes <- rabi::rs_IDs(total.length, redundancy, alphabet)
 #prep the output to be displayed in a table
codes <- t(do.call("cbind",codes))
knitr::kable(
  head(codes, n = 10L), 
  col.names = c("Head","Thorax","R. Abdomen","L. Abdomen"), 
  align = "c", caption = "ID sequences for ants")
```

| Head | Thorax | R. Abdomen | L. Abdomen |
| :--: | :----: | :--------: | :--------: |
|  0   |   0    |     0      |     0      |
|  1   |   1    |     1      |     1      |
|  2   |   2    |     2      |     2      |
|  3   |   3    |     3      |     3      |
|  4   |   4    |     4      |     4      |
|  0   |   1    |     2      |     3      |
|  1   |   2    |     3      |     4      |
|  2   |   3    |     4      |     0      |
|  3   |   4    |     0      |     1      |
|  4   |   0    |     1      |     2      |

ID sequences for ants

We can easily add our paint color names to the output as well. (However,
I would personally save the mapping for future reference, etc.)

``` r
paint.names <- c("red","light blue", "greenish", "off-white", "yellow")
color.codes <- rabi::codes_to_colors(codes, paint.names)
 #prep the output to be displayed in a table
color.codes <- t(do.call("cbind",color.codes))
knitr::kable(
  head(color.codes, n = 10L), 
  col.names = c("Head","Thorax","R. Abdomen","L. Abdomen"), 
  align = "c", caption = "ID paint sequences for ants")
```

|    Head    |   Thorax   | R. Abdomen | L. Abdomen |
| :--------: | :--------: | :--------: | :--------: |
|    red     |    red     |    red     |    red     |
| light blue | light blue | light blue | light blue |
|  greenish  |  greenish  |  greenish  |  greenish  |
| off-white  | off-white  | off-white  | off-white  |
|   yellow   |   yellow   |   yellow   |   yellow   |
|    red     | light blue |  greenish  | off-white  |
| light blue |  greenish  | off-white  |   yellow   |
|  greenish  | off-white  |   yellow   |    red     |
| off-white  |   yellow   |    red     | light blue |
|   yellow   |    red     | light blue |  greenish  |

ID paint sequences for
    ants

    #> Note: The mapping (see below) that was used to assign color names to numeric values is not saved or assigned to a variable. 
    #>  The exact mapping may change with repeated function calls. Depending on your circumstances, you may want to record this now.

    #>            0            1            2            3            4 
    #>        "red" "light blue"   "greenish"  "off-white"     "yellow"

\#\#More For a (possibly buggy) Shiny-based GUI suited for new R users,
try running this command.

``` r
rabi::exampleGUI()
```

For additional functions, deeper examples, and such, check out the
vignette on designing color band schemes for bird legs: it’s chock full
of cool stuff.

``` r
utils::vignette('loosebirdtag', package='rabi')
```

Lastly, check out our journal article on this subject:

> Burchill, A. T., & Pavlic, T. P. (2019). Dude, where’s my mark?
> Creating robust animal identification schemes informed by
> communication theory. *Animal Behaviour*, 154, 203-208.
> [doi:10.1016/j.anbehav.2019.05.013](https://doi.org/10.1016/j.anbehav.2019.05.013)


NEWS.md
rabi 1.0.2 (2019-12-08)
=========================

### Bug fixes

  * changed class(.)==* instances to is(.,*) for compatibility
  * added CITATION reference to published paper in Animal Behaviour journal
  


rabi 1.0.0 (2018-01-17)
=========================

### NEW FEATURES

  * added a Shiny-based GUI for the rs_IDs() function, exampleGUI()
  

rabi 0.1.2 (2017-10-17)
=========================

### Bug fixes

  * fixed the deprecated function names in the vignette to match current names

rabi 0.1.0 (2017-10-17)
=========================

### NEW FEATURES

  * being hopefully submitted to CRAN!


R/brutegenerator.R
#' Brute force color coding scheme generator
#'
#' Generates "color" coding schemes used to mark and identify individual animals. The codes are robust to an arbitrary number of partial erasures. This method uses a sloppy, very slow, stochastic brute force method.
#'
#'
#' @param total.length the number of unique positions to be marked on the animal. (This can be thought of as the total number of positions on which color bands or paint marks will be applied.)
#' @param redundancy the number of erasures that can occur without disrupting surety of unique identification. This value determines how robust the scheme is to erasures.
#' @param alphabet an integer representing the 'alphabet size.' This is the number of unique markings (think different paint colors, symbols, or varieties of bands) at your disposal.Note: unlike the Reed-Solomon inspired function, \code{\link{rs_IDs}}, this function can take non-prime values.
#' @param num.tries the number of iterations that will be run before choosing the best option. Increasing this number increases the running time.
#' @param available.colors an optional list of strings that contains the names of the unique markings which compose the given 'alphabet' (e.g. "blue", "red", "yellow", etc.). If left blank, the mapping can be done at any later time using \code{\link{codes_to_colors}}. Additionally, the length of this list must match the 'alphabet size' given above.
#'
#' @details This function generates the list of all possible unique ID codes given the 'alphabet size' (\code{alphabet}) and the number of unique positions available for marking (\code{total.length}). The list of combinations is then iteratively pruned down until the required robustness has been reached; the \href{https://en.wikipedia.org/wiki/Hamming_distance}{distance} between any two ID codes must greater than the value specified in \code{redundancy}.
#'
#' However, the iterative pruning is done randomly, so it is likely that resulting list of codes does not contain the maximum possible number of robust codes. Thus, the process is repeated multiple times (\code{num.tries}) and the list that contains the largest number of robust codes is kept and returned.
#'
#'
#' @return a list of unique ID codes that fit the provided parameters.
#'
#' If an appropriate argument for \code{available.colors} is provided, each code will be a sequence of strings, otherwise, each code will be a sequence of numeric values.
#'
#' @note  the \code{\link{rs_IDs}} function always generates the maximum number of unique codes per scheme. However, \code{\link{rs_IDs}} suffers from certain limitations that \code{\link{brute_IDs}} does not: it requires \code{alphabet} to be a prime number, \code{total.length} to be less than or equal to \code{alphabet}, etc.
#'
#' @author Andrew Burchill, \email{andrew.burchill@asu.edu}
#' @references Burchill, A. T., & Pavlic, T. P. (2019). Dude, where's my mark? Creating robust animal identification schemes informed by communication theory. \emph{Animal Behaviour}, 154, 203-208. \href{https://doi.org/10.1016/j.anbehav.2019.05.013}{doi:10.1016/j.anbehav.2019.05.013}
#' @seealso \code{\link{rs_IDs}}, \code{\link{tweaked_IDs}}, \code{\link{simple_IDs}}. Also see the vignette \href{../doc/loosebirdtag.html}{\code{loosebirdtag}} for demonstrations and additional uses.
#'
#' @examples
#' total.length <- 6  #we have six positions to mark,
#' redundancy <- 2    #we want surety even with two erasures,
#' alphabet <- 4      #and we currently have five types of paint in stock
#'
#'  #This gives a warning because rs_IDs() only accepts
#'  #a pretty constrained set of parameters
#' codes <- rs_IDs(total.length, redundancy, alphabet)
#' length(codes)
#'
#'  #However, we can do it with brute_IDs() to get more unique IDs
#' codes <- brute_IDs(total.length, redundancy, alphabet, num.tries = 1)
#' length(codes)
#'
#'  #Let's make those into human-readable color sequences
#' color.names <- c("blue","red","green","pink","yellow-with-a-stripe")
#' color.codes <- codes_to_colors(codes, color.names)
#'
#'
#'
#' @export
#' @importFrom stringdist seq_distmatrix
#' @importFrom methods is

brute_IDs <- function(total.length, redundancy, alphabet, num.tries = 10, available.colors = NULL) {

  if (missing(alphabet)) {
    stop("Error: you need to enter an 'alphabet size,' e.g. the number of paint colors you have")
  }
  if (missing(total.length)) {
    stop("Error: you need to enter the total length of the ID, e.g. how many color bands or paint drops on each organism")
  }
  if (missing(redundancy)) {
    stop("Error: you need specify to how many erasure events the IDs should be robust. Note, an increase in robustness requires an increase in the total length of the ID. ")
  }
  if (redundancy >= total.length || redundancy == 0) {
    stop("Error: the code must be robust to at least one erasure. It also cannot be robust to a number of positions equal to or greater than the total length.")
  }
  if (!is(num.tries, "numeric")) {
    stop(paste0("Error: the variable 'num.tries' must be of the class 'numeric,' not '", class(num.tries),".'"))
  }

  tester <- function(total.length, redundancy, alphabet) {
     #generate all sequences and turn into a list
    perms <- rep(list(seq_len(alphabet)),total.length)
    combos <- as.matrix(expand.grid(perms)) - 1
    combo.list <- split(combos, 1:nrow(combos))
    names(combo.list) <- NULL
     #pick a random sequence and start making the "safe" list with it
    x <- sample(1:length(combo.list), 1)
    new.combs <- combo.list[x]
    names(new.combs) <- NULL
     #remove everything too similar to the chosen sequence from the old list
    combo.list <- combo.list[stringdist::seq_distmatrix(combo.list, new.combs, method = "hamming")[, length(new.combs)] > redundancy]
    names(combo.list) <- 1:length(combo.list)
     #do this again and again until everything is removed
    while (length(combo.list) > 0) {
      x <- sample(1:length(combo.list), 1)
      new.combs[length(new.combs) + 1] <- (combo.list[x])
      combo.list <- combo.list[stringdist::seq_distmatrix(combo.list, new.combs, method = "hamming")[, length(new.combs)] > redundancy]
      if (length(combo.list) != 0) {
        names(combo.list) <- 1:length(combo.list)
      }
    }
    # print(length(new.combs)) table(unlist(seq_distmatrix(new.combs,new.combs,method='hamming')))
    return(new.combs)
  }
   #run through the function several times and keep the best
  temp1 <- NULL
  temp2 <- 0
  for (i in 1:num.tries) {
    temp1 <- invisible(tester(total.length, redundancy, alphabet))
    if (length(temp1) > length(temp2))
      temp2 <- temp1
  }
  temp2 <- codes_to_colors(temp2, available.colors)
  return(temp2)
}
